opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py

```diff
@@ -114,9 +128,113 @@ def on_end(self, span: "Span") -> None:
+                return False
+
+        return True
+
+
+class ConcurrentMultiSpanProcessor(SpanProcessor):
+    """Implementation of :class:`SpanProcessor` that forwards all received
+    events to a list of `SpanProcessor`s in parallel.
+
+    Calls to the underlying `SpanProcessor`s are forwarded in parallel by
+    submitting them to a thread pool executor and waiting until each span
+    processor finished its work.
+
+    Args:
+        num_threads: The number of threads managed by the thread pool executor
+            and thus defining how many span processors can work in parallel.
+    """
+
+    def __init__(self, num_threads: int = 2):
+        # use a tuple to avoid race conditions when adding a new span and
+        # iterating through it on "on_start" and "on_end".
+        self._span_processors = ()  # type: Tuple[SpanProcessor, ...]
+        self._lock = threading.Lock()
+        self._executor = concurrent.futures.ThreadPoolExecutor(
+            max_workers=num_threads
+        )
+
+    def add_span_processor(self, span_processor: SpanProcessor) -> None:
+        """Adds a SpanProcessor to the list handled by this instance."""
+        with self._lock:
+            self._span_processors = self._span_processors + (span_processor,)
+
+    def _submit_and_await(
+        self, func: Callable[[SpanProcessor], Callable[..., None]], *args: Any
+    ):
+        futures = []
+        for sp in self._span_processors:
+            future = self._executor.submit(func(sp), *args)
+            futures.append(future)
+        for future in futures:
+            future.result()
+
+    def on_start(self, span: "Span") -> None:
+        self._submit_and_await(lambda sp: sp.on_start, span)
+
+    def on_end(self, span: "Span") -> None:
+        self._submit_and_await(lambda sp: sp.on_end, span)
+
+    def shutdown(self) -> None:
+        """Shuts down all underlying :class:`SpanProcessor`s in parallel.
+        """
+        self._submit_and_await(lambda sp: sp.shutdown)
+
+    def force_flush(self, timeout_millis: int = 30000) -> bool:
+        """Calls force_flush on all underlying :class:`SpanProcessor`s in
+        parallel.
+
+        Args:
+            timeout_millis: The maximum amount of time to wait for spans to be
+                exported.
+
+        Returns:
+            True if all `SpanProcessor`s flushed their spans within the given
+            timeout, False otherwise.
+        """
+        futures = []
+        for sp in self._span_processors:  # type: SpanProcessor
+            future = self._executor.submit(sp.force_flush, timeout_millis)
+            futures.append(future)
+        deadline_ns = time_ns() + timeout_millis * 1000000
+        flushed_in_time = True
+        for future in futures:
+            try:
+                timeout_sec = (deadline_ns - time_ns()) / 1e9
+                flushed_in_time = (
+                    future.result(timeout_sec) and flushed_in_time
```

Comment:
I don't believe this will have the intended effect.

Future.result will block and time out for up to timeout_sec per call. As a result, if you have 2 futures, you'll actually be blocked for up to n * timeout_sec:

1. future.result(timeout_sec)
2. doesn't timeout, but finishes just under timeout_sec (~ timeout_sec time has passed)
3. future.result(timeout_sec) #2 
4. doesn't timoeut, but finishes just under timeout_sec (~timeout_sec * 2 time has passed)

and so on.

I believe map may be a more appropriate choice here: https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Executor.map

Reply:
nvm, I see the timeout_sec is calculated every round, I believe this *will* work, but you could simplify the logic significantly by just passing all the futures to a wait function: https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.wait